### PR TITLE
Add support to read&wrtite Grail segment-feature type of configuration

### PR DESCRIPTION
**Pull request by jskelin**
#### What this PR does / Why we need it:
Added support to read and write the Grail segment-filter configuration. To use it, the `MONACO_FILTER_SEGMENT` environment variable needs to be present.

### CODE REVIEW

File: pkg/persistence/config/writer/config_writer_test.go [hunk 1]
```diff
@@ -1023,6 +1026,47 @@ func TestWriteConfigs(t *testing.T) {
 				"project/bucket/mybucket.json",
 			},
 		},
+		{
+			name:    "Grail file-segment",
```
Comment (arthurpitman) [inline pkg/persistence/config/writer/config_writer_test.go:1030]:
```suggestion
			name:    "Grail filter segment",
```

File: pkg/config/config.go [hunk 2]
```diff
@@ -64,102 +65,11 @@ var ReservedParameterNames = []string{IdParameter, NameParameter, ScopeParameter
...
-	ClassicApiTypeId   TypeId = "classic"
-	EntityTypeId       TypeId = "entity"
-	AutomationTypeId   TypeId = "automation"
-	BucketTypeId       TypeId = "bucket"
-	DocumentTypeId     TypeId = "document"
-	OpenPipelineTypeId TypeId = "openpipeline"
-)
-
-type Type interface {
+type Typer interface {
```
[the remark this answers is not in the document]
Comment (Laubi) [reply pkg/config/config.go:70]: Agree as well that `Type` not straightforward, but we should discuss together what we sould do here. I don't think renaming this here and now is what we should work on.

File: pkg/config/types.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+const (
+	SettingsTypeID       TypeID = "settings"
```
[the remark this answers is not in the document]
Comment (Laubi) [reply pkg/config/types.go:20]: This functionality has just moved from `config.go` to here. If I see it correctly; the `ID` method was already here before. 
I do see the moving the code to this file to have the functionality a bit split.

File: internal/featureflags/temporary.go [hunk 4]
```diff
@@ -38,6 +38,9 @@ const (
 	// from the dependency graph created by Monaco. These configs are not only skipped during deployment but also
 	// not validated prior to deployment. Further, other configs cannot reference properties of this config anymore.
 	IgnoreSkippedConfigs TemporaryFlag = "MONACO_FEAT_IGNORE_SKIPPED_CONFIGS"
+	// GrailFilterSegment toggles whether filter-segment configurations are downloaded and / or deployed.
+	// Introduced: v2.18.0
+	GrailFilterSegment TemporaryFlag = "MONACO_FILTER_SEGMENT"
```
Comment (Laubi) [inline internal/featureflags/temporary.go:43]: We should get rid of the `Grail` wording and just call it `FilterSegment`

File: pkg/persistence/config/internal/persistence/type_definition.go [hunk 5]
```diff
@@ -322,7 +334,10 @@ func (c TypeDefinition) MarshalYAML() (interface{}, error) {
 				},
 			}, nil
 		}
-	}
 
+	case config.GrailFilterSegment:
+		return GrailFilterSegment, nil
```
Comment (Laubi) [inline pkg/persistence/config/internal/persistence/type_definition.go:339]: The cases above also check the FF, please do it here as well

File: pkg/persistence/config/loader/config_loader_test.go [hunk 6]
```diff
@@ -1167,6 +1169,48 @@ configs:
 `,
 			wantErrorsContain: []string{"unknown API: bucket"},
 		},
+		{
+			name:             "grail filter-segment config with FF on",
+			envVars:          map[string]string{featureflags.Temporary[featureflags.GrailFilterSegment].EnvName(): "true"},
```
Comment (Laubi) [inline pkg/persistence/config/loader/config_loader_test.go:1174]: Please add a case with FF off

File: pkg/persistence/config/loader/config_loader_test.go [hunk 7]
```diff
@@ -1167,6 +1169,48 @@ configs:
...
+			wantConfigs: []config.Config{
+				{
+					Coordinate: coordinate.Coordinate{
+						Project:  "project",
+						Type:     "filter-segment",
+						ConfigId: "profile-id",
+					},
+					Type:        config.GrailFilterSegment{},
+					Template:    template.NewInMemoryTemplate("profile.json", "{}"),
+					Parameters:  config.Parameters{},
```
Comment (Laubi) [inline pkg/persistence/config/loader/config_loader_test.go:1193]:
```suggestion
					Template:    template.NewInMemoryTemplate("filter-segment.json", "{}"),
```

File: pkg/persistence/config/writer/config_writer_test.go (hunk 1 above)
Comment (Laubi) [reply pkg/persistence/config/writer/config_writer_test.go:1030]:
```suggestion
			name:    "Filter-Segments",
```

File: pkg/persistence/config/writer/config_writer_test.go [hunk 8]
```diff
@@ -1023,6 +1026,47 @@ func TestWriteConfigs(t *testing.T) {
...
+			envVars: map[string]string{featureflags.Temporary[featureflags.GrailFilterSegment].EnvName(): "true"},
+			configs: []config.Config{
```
Comment (Laubi) [inline pkg/persistence/config/writer/config_writer_test.go:1032]: I'd also add here a test with the FF off that an error occurs

File: pkg/config/config.go [hunk 9]
```diff
@@ -64,102 +65,11 @@ var ReservedParameterNames = []string{IdParameter, NameParameter, ScopeParameter
 // Parameters defines a map of name to parameter
 type Parameters map[string]parameter.Parameter
 
-type TypeId string
+type TypeID string
```
[the remark this answers is not in the document]
Comment (jskelin) [reply pkg/config/config.go:68]: This renaming is to satisfy GO recommendation - https://google.github.io/styleguide/go/decisions#initialisms.
I agree, next time it'll be in separate PR

File: internal/featureflags/temporary.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jskelin) [reply internal/featureflags/temporary.go:43]: done

File: pkg/persistence/config/writer/config_writer_test.go (hunk 1 above)
Comment (jskelin) [reply pkg/persistence/config/writer/config_writer_test.go:1030]: I would keep Grail as it is in the documentation under Grail. Also the name of the Open API spec is Grail Filter-Segment

File: pkg/persistence/config/writer/config_writer_test.go (hunk 8 above)
Comment (jskelin) [reply pkg/persistence/config/writer/config_writer_test.go:1032]: Good suggestion.

File: pkg/persistence/config/internal/persistence/type_definition.go (hunk 5 above)
Comment (jskelin) [reply pkg/persistence/config/internal/persistence/type_definition.go]: added

File: pkg/persistence/config/loader/config_loader_test.go (hunk 6 above)
Comment (jskelin) [reply pkg/persistence/config/loader/config_loader_test.go:1174]: added

File: pkg/persistence/config/loader/config_loader_test.go (hunk 7 above)
Comment (jskelin) [reply pkg/persistence/config/loader/config_loader_test.go:1193]: I don't understand this suggestion. `Template` is already set in line 1192

File: pkg/config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jskelin) [reply pkg/config/config.go]: renaming revoekd

File: pkg/config/types.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	ClassicApiTypeID     TypeID = "classic"
+	EntityTypeID         TypeID = "entity"
+	AutomationTypeID     TypeID = "automation"
+	BucketTypeID         TypeID = "bucket"
+	DocumentTypeID       TypeID = "document"
+	OpenPipelineTypeID   TypeID = "openpipeline"
+	GrailFilterSegmentID TypeID = "filter-segment"
+)
+
+var _ Typer = SettingsType{}
```
[the remark this answers is not in the document]
Comment (jskelin) [reply pkg/config/types.go]: rename revoked

File: pkg/persistence/config/loader/config_loader_test.go (hunk 7 above)
Comment (jskelin) [reply pkg/persistence/config/loader/config_loader_test.go:1193]: It is about renaming the name of the file...

**Review by Laubi — APPROVED** [review]
Approved. However, we need to fix the naming in a follow-up PR as soon as we have the final name we should use.

Requirement: This PR **must** be merged using Rebase on main to keep the commits separate.

**Comment by jskelin** [comment]
> Approved. However, we need to fix the naming in a follow-up PR as soon as we have the final name we should use.
> 
> Requirement: This PR **must** be merged using Rebase on main to keep the commits separate.

Agree.

**Pull request merged**